tools/workload-simulator.html

```diff
@@ -0,0 +1,1160 @@
+          multisample: { count: multisampling.checked ? sampleCount : 1, },
+          vertex: {
+            entryPoint: 'vs',
+            module: shaderModule,
+          },
+          fragment: {
+            entryPoint: 'fs',
+            targets: [ { format: swapChainFormat, }, ],
+            module: shaderModule,
+          },
+        });
+        sampler = device.createSampler({
+          magFilter: 'linear',
+          minFilter: 'linear',
+        });
+        const texture = device.createTexture({
+          size: [256, 256, 1],
+          format: 'rgba8unorm',
+          usage: GPUTextureUsage.SAMPLED | GPUTextureUsage.COPY_DST,
+        });
+        const loadImage = async () => {
+          const imageBitmap = await createImageBitmap(img);
+          device.queue.copyImageBitmapToTexture(
+            { imageBitmap }, { texture }, [256, 256, 1]);
+          render();
+        };
+        if (img.complete) loadImage(); else img.addEventListener('load', loadImage);
+        uniformBuffer = device.createBuffer({
+          size: 2 * 4,
+          usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
+        });
+        bindGroup = device.createBindGroup({
+          layout: bindGroupLayout,
+          entries: [
+            { binding: 0, resource: { buffer: uniformBuffer } },
+            { binding: 1, resource: sampler },
+            { binding: 2, resource: texture.createView() },
+          ],
+        });
+      } catch(e) {
+        errorMessage.textContent = "Error initializing WebGPU: " + e;
+        return;
+      }
+    }
+
+    // End WebGPU init.
+    // Begin WebGPU rendering.
+
+    if (!device) {
+      errorMessage.textContent = "Error: WebGPU not supported.";
+      return;
+    }
+    try {
+      const swapChainView = swapChain.getCurrentTexture().createView();
+      const commandBuffers = [];
+      device.queue.writeBuffer(uniformBuffer, 0, new Float32Array([position[0]/size*2 - 0.5, -position[1]*2/size + 0.5]));
+      let buffer = null;
+
+      // Upload data using createBuffer with mappedAtCreation.
+      if (bufferData.value > 0) {
+        buffer = device.createBuffer({ size: bufferDataArray.byteLength, usage: GPUBufferUsage.VERTEX, mappedAtCreation: true });
+        new Float32Array(buffer.getMappedRange()).set(bufferDataArray);
+        buffer.unmap();
+        // TODO: use this data in rendering so it can't be optimized away
+      }
+
+      // Upload data using queue.writeBuffer.
+      if (bufferSubData.value > 0) {
+        if (!uploadBuffer || uploadBufferSize != bufferSubDataArray.byteLength) {
+          if (uploadBuffer) uploadBuffer.destroy();
+          uploadBufferSize = bufferSubDataArray.byteLength;
+          uploadBuffer = device.createBuffer({
+            size: bufferSubDataArray.byteLength,
+            usage: GPUBufferUsage.VERTEX | GPUBufferUsage.COPY_DST,
+          });
+        }
+        device.queue.writeBuffer(uploadBuffer, 0, bufferSubDataArray);
```

Comment:
This `uploadBuffer` isn't used for anything?

Reply:
Yes, it's just to test upload performance. Ideally the data would be used in rendering so a sufficiently smart implementation couldn't optimize away the upload. I doubt that's a problem currently.